File: server/src/main/java/org/jetbrains/bsp/bazel/server/bep/BepOutputBuilder.java
package org.jetbrains.bsp.bazel.server.bep;

import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.BuildEventId;
import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.NamedSetOfFiles;
import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.OutputGroup;
import java.net.URI;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

class BepOutputBuilder {
  private final Map<String, Set<String>> outputGroups = new HashMap<>();
  private final Map<String, TextProtoDepSet> textProtoFileSets = new HashMap<>();
  private final Set<String> rootTargets = new HashSet<>();

  public void storeNamedSet(String id, NamedSetOfFiles namedSetOfFiles) {
    this.textProtoFileSets.put(
        id,
        new TextProtoDepSet(
            namedSetOfFiles.getFilesList().stream()
                .filter(it -> it.getName().endsWith("bsp-info.textproto"))
                .map(it -> URI.create(it.getUri()))
                .collect(Collectors.toList()),
            namedSetOfFiles.getFileSetsList().stream()
                .map(it -> it.getId())
                .collect(Collectors.toList())));
  }

  public void storeTargetOutputGroups(String target, List<OutputGroup> outputGroups) {
    rootTargets.add(target);

    for (var group : outputGroups) {
      var fileSets =
          group.getFileSetsList().stream()
              .map(BuildEventId.NamedSetOfFilesId::getId)
              .collect(Collectors.toList());
      this.outputGroups.computeIfAbsent(group.getName(), key -> new HashSet<>()).addAll(fileSets);
    }
  }

  public BepOutput build() {
    return new BepOutput(outputGroups, textProtoFileSets, rootTargets);
  }

  private final Logger LOGGER = LogManager.getLogger(BepOutputBuilder.class);
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/bep/BepOutputBuilder.kt
package org.jetbrains.bsp.bazel.server.bep

import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos
import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.NamedSetOfFiles
import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.OutputGroup
import org.jetbrains.bsp.bazel.server.paths.BazelPathsResolver
import java.io.File
import java.nio.file.Path
import kotlin.io.path.Path

data class TextProtoDepSet(val files: Collection<Path>, val children: Collection<String>)

internal class BepOutputBuilder(private val bazelPathsResolver: BazelPathsResolver) {
  private val outputGroups: MutableMap<String, MutableSet<String>> = HashMap()
  private val textProtoFileSets: MutableMap<String, TextProtoDepSet> = HashMap()
  private val rootTargets: MutableSet<String> = HashSet()

  fun storeNamedSet(id: String, namedSetOfFiles: NamedSetOfFiles) {
    val textProtoDepSet = TextProtoDepSet(
      files = namedSetOfFiles
        .filesList
        .filter { it.name.endsWith("bsp-info.textproto") }
        .map { it.toLocalPath() },
      children = namedSetOfFiles.fileSetsList.map { it.id }
    )

    textProtoFileSets[id] = textProtoDepSet
  }

  private fun BuildEventStreamProtos.File.toLocalPath(): Path {
    val mergedPathPrefix = Path(pathPrefixList.joinToString(File.separator))
    val bazelOutputRelativePath = mergedPathPrefix.resolve(name)

    return bazelPathsResolver.resolveOutput(bazelOutputRelativePath)
  }

  fun storeTargetOutputGroups(target: String, outputGroups: List<OutputGroup>) {
    rootTargets.add(target)

    for (group in outputGroups) {
      val fileSets = group.fileSetsList.map { it.id }
      this.outputGroups.computeIfAbsent(group.name) { HashSet() }.addAll(fileSets)
    }
  }

  fun clear() {
    outputGroups.clear()
    textProtoFileSets.clear()
    rootTargets.clear()
  }

  fun build(): BepOutput = BepOutput(outputGroups, textProtoFileSets, rootTargets)
}
